File: Pandacap.LowLevel/Data/TheirContent/ATProto/BlueskyFeedItem.fs
namespace Pandacap.Data

open System
open System.ComponentModel.DataAnnotations
open Pandacap.Html
open Pandacap.PlatformBadges

type BlueskyFeedItemUser() =
    member val DID = "" with get, set
    member val PDS = nullString with get, set
    member val DisplayName = nullString with get, set
    member val Handle = "" with get, set
    member val Avatar = nullString with get, set

type BlueskyFeedItemImage() =
    member val Thumb = "" with get, set
    member val Fullsize = "" with get, set
    member val Alt = "" with get, set

type BlueskyFeedItem() =
    [<Key>]
    member val Id = Guid.Empty with get, set

    member val CID = "" with get, set
    member val RecordKey = "" with get, set
    member val Author = new BlueskyFeedItemUser() with get, set
    member val PostedBy = new BlueskyFeedItemUser() with get, set
    member val CreatedAt = DateTimeOffset.MinValue with get, set
    member val IndexedAt = DateTimeOffset.MinValue with get, set
    member val IsAdultContent = false with get, set
    member val Text = "" with get, set
    member val Images = new ResizeArray<BlueskyFeedItemImage>() with get, set
    member val DismissedAt = nullDateTimeOffset with get, set

    interface IBlueskyPost with
        member this.CID = this.CID
        member this.DID = this.Author.DID
        member this.PDS = this.Author.PDS
        member this.RecordKey = this.RecordKey
        member _.InFavorites = false

    interface IInboxPost with
        member this.DismissedAt
            with get () = this.DismissedAt
             and set value = this.DismissedAt <- value
        member _.IsPodcast = false
        member this.IsShare = this.PostedBy.DID <> this.Author.DID

    interface IPost with
        member _.Platform = Bluesky
        member this.Url = $"https://{this.PostedBy.PDS}"
        member this.DisplayTitle =
            TitleGenerator.FromBody this.Text
            |> ExcerptGenerator.FromText 60
        member this.Id = $"{this.Id}"
        member this.LinkUrl = $"/ATProto/ViewKnownBlueskyPost?id={this.Id}"
        member this.PostedAt = this.CreatedAt
        member this.ProfileUrl = $"https://bsky.app/profile/{this.PostedBy.DID}"
        member this.Thumbnails = [
            if not this.IsAdultContent then
                for image in this.Images do {
                    new IPostThumbnail with
                        member _.AltText = image.Alt
                        member _.Url = image.Thumb
                }
        ]
        member this.Usericon = this.PostedBy.Avatar
        member this.Username = this.PostedBy.Handle

File: Pandacap.Database/BlueskyFeedItem.cs
using Pandacap.Text;
using Pandacap.UI.Badges;
using Pandacap.UI.Elements;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Security.Cryptography;

namespace Pandacap.Database
{
    public abstract class BlueskyFeedItem : IInboxPost
    {
        [Key]
        public string CID { get; set; } = "";

        public DateTimeOffset CreatedAt { get; set; }

        public List<string> Labels { get; set; } = [];

        public string Text { get; set; } = "";

        public class Image
        {
            public string CID { get; set; } = "";
            public string Alt { get; set; } = "";
        }

        public List<Image> Images { get; set; } = [];

        public DateTimeOffset? DismissedAt { get; set; }

        [NotMapped]
        public bool AdultContent => Labels
            .Intersect(["porn", "sexual", "nudity", "sexual-figurative", "graphic-media"])
            .Any();

        public class User
        {
            public string DID { get; set; } = "";
            public string PDS { get; set; } = "";
            public string? DisplayName { get; set; }
            public string? Handle { get; set; } = "";
            public string? AvatarCID { get; set; }
        }

        [NotMapped]
        public abstract string OriginalDID { get; }

        [NotMapped]
        public abstract string OriginalPDS { get; }

        [NotMapped]
        public abstract string OriginalRecordKey { get; }

        [NotMapped]
        public abstract User AttributeTo { get; }

        [NotMapped]
        public abstract DateTimeOffset DateTo { get; }

        [NotMapped]
        public abstract bool IsShare { get; }

        bool IInboxPost.IsPodcast => false;

        Badge IPost.Badge => Badges.ATProto.WithText(OriginalPDS);

        string IPost.DisplayTitle => ExcerptGenerator.FromText(60, Text);

        string IPost.Id => $"{CID}";

        string IPost.InternalUrl => $"/ATProto/ViewBlueskyPost?did={OriginalDID}&rkey={OriginalRecordKey}";

        string IPost.ExternalUrl => $"https://bsky.app/profile/{OriginalDID}/post/{OriginalRecordKey}";

        DateTimeOffset IPost.PostedAt => DateTo;

        string? IPost.ProfileUrl => $"https://bsky.app/profile/{AttributeTo.DID}";

        IEnumerable<IPostThumbnail> IPost.Thumbnails => Images.Select(image =>
            !AdultContent
            ? new PostThumbnail(
                $"/ATProto/GetBlob?did={OriginalDID}&cid={image.CID}",
                image.Alt)
            : new PostThumbnail(
                "/images/trgray.svg",
                ""));

        string? IPost.Username => AttributeTo.Handle ?? AttributeTo.DID;

        string? IPost.Usericon => AttributeTo.AvatarCID is string cid
            ? $"/ATProto/GetBlob?did={AttributeTo.DID}&cid={cid}"
            : null;

        private class PostThumbnail(string url, string alt) : IPostThumbnail
        {
            string IPostThumbnail.Url => url;

            string IPostThumbnail.AltText => alt;
        }

        private class HiddenPostThumbnail : IPostThumbnail
        {
            string IPostThumbnail.Url => "/images/trgray.svg";

            string IPostThumbnail.AltText => "";
        }
    }
}
